clients/src/main/java/org/apache/kafka/common/record/LogBufferIterator.java

```diff
@@ -0,0 +1,219 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.CorruptRecordException;
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.common.utils.Utils;
+
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+
+/**
+ * An iterator which handles both the shallow and deep iteration of a log buffer.
+ */
+public class LogBufferIterator extends AbstractIterator<LogEntry> {
+    private final boolean shallow;
+    private final boolean ensureMatchingMagic;
+    private final int maxMessageSize;
+    private final ShallowRecordsIterator<?> shallowIter;
+    private DeepRecordsIterator innerIter;
+
+    public LogBufferIterator(LogInputStream<?> logInputStream,
+                             boolean shallow,
+                             boolean ensureMatchingMagic,
+                             int maxMessageSize) {
+        this.shallowIter = new ShallowRecordsIterator<>(logInputStream);
+        this.shallow = shallow;
+        this.ensureMatchingMagic = ensureMatchingMagic;
+        this.maxMessageSize = maxMessageSize;
+    }
+
+    /**
+     * Get a shallow iterator over the given input stream.
+     * @param logInputStream The log input stream to read the entries from
+     * @param <T> The type of the log entry
+     * @return The shallow iterator.
+     */
+    public static <T extends LogEntry> Iterator<T> shallowIterator(LogInputStream<T> logInputStream) {
+        return new ShallowRecordsIterator<>(logInputStream);
+    }
+
+    /*
+     * Read the next record from the buffer.
+     *
+     * Note that in the compressed message set, each message value size is set as the size of the un-compressed
+     * version of the message value, so when we do de-compression allocating an array of the specified size for
+     * reading compressed value data is sufficient.
```

Comment:
To be honest, I'm not really sure why this comment is needed. It seems obvious that the key and value sizes in the inner messages are based on the uncompressed data (how could they be otherwise if we compress the inner message set as a whole?).